From 9e4d2874690eabac8c4b1a4909044d728aa2095d Mon Sep 17 00:00:00 2001
From: Francesco VIRLINZI <francesco.virlinzi@st.com>
Date: Tue, 23 Jun 2009 15:39:38 +0200
Subject: [PATCH] sh_stm: pm: Fixed pm debug support in the stx7141 [v3]

This patch fixes the PM debug support in the stx7141.
When CONFIG_PM_DEBUG is enabled the system asks for a PIO
normally used by SSC, creating a fault during the boot.

Signed-off-by: Francesco Virlinzi <francesco.virlinzi@st.com>
Signed-off-by: Stuart Menefy <stuart.menefy@st.com>
---
 arch/sh/kernel/cpu/sh4/suspend-stx7141.c |   21 ++++++++++++++++-----
 1 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/arch/sh/kernel/cpu/sh4/suspend-stx7141.c b/arch/sh/kernel/cpu/sh4/suspend-stx7141.c
index 3dffc3a..3c5898a 100644
--- a/arch/sh/kernel/cpu/sh4/suspend-stx7141.c
+++ b/arch/sh/kernel/cpu/sh4/suspend-stx7141.c
@@ -193,12 +193,23 @@ static struct sh4_suspend_t st40data __cacheline_aligned = {
 static int __init suspend_platform_setup()
 {
 	struct sysconf_field *sc;
-#ifdef CONFIG_PM_DEBUG
+#if defined(CONFIG_PM_DEBUG)
+	struct stpio_pin *pin;
 	/* route the sh4/2  clock frequenfy */
-	iowrite32(0xc, CLOCKGENA_BASE_ADDR + CKGA_CLKOBS_MUX1_CFG);
-	stpio_request_set_pin(3, 2, "clkA dbg", STPIO_ALT_OUT, 1);
-	sc = sysconf_claim(SYS_CFG, 19, 22, 23, "clkA dbg");
-	sysconf_write(sc, 11);
+	pin = stpio_request_set_pin(3, 2, "clkA dbg", STPIO_ALT_OUT, 1);
+	if (pin) {
+		iowrite32(0xc, CLOCKGENA_BASE_ADDR + CKGA_CLKOBS_MUX1_CFG);
+		sc = sysconf_claim(SYS_CFG, 19, 22, 23, "clkA dbg");
+		if (sc)
+			sysconf_write(sc, 11);
+		else {
+			printk(KERN_ERR
+			   "Not able to acquire the sysconf 19 [22,23]\n");
+			stpio_free_pin(pin);
+		}
+	} else
+		printk(KERN_ERR
+			"Not able to route the ClkA on external pin\n");
 #endif
 
 	sc = sysconf_claim(SYS_CFG, 38, 20, 20, "pm");
-- 
1.6.0.6

